File: src/main/java/com/meltwater/puppy/rest/RestRequestBuilder.java
package com.meltwater.puppy.rest;

import com.google.common.escape.Escaper;
import com.google.common.net.UrlEscapers;
import org.glassfish.jersey.client.JerseyClient;
import org.glassfish.jersey.client.JerseyClientBuilder;
import org.glassfish.jersey.client.JerseyInvocation;
import org.glassfish.jersey.client.JerseyWebTarget;
import org.glassfish.jersey.client.authentication.HttpAuthenticationFeature;

import java.util.HashMap;
import java.util.Map;

public class RestRequestBuilder {

    private String host = null;
    private String authUser = null;
    private String authPass = null;
    private String authUserNext = null;
    private String authPassNext = null;
    private Map<String, String> headers = new HashMap<>();

    private final JerseyClient client = JerseyClientBuilder.createClient();
    private final Escaper escaper = UrlEscapers.urlPathSegmentEscaper();

    public RestRequestBuilder() {
    }

    public RestRequestBuilder withHost(String host) {
        this.host = host;
        return this;
    }

    public RestRequestBuilder withAuthentication(String authUser, String authPass) {
        this.authUser = authUser;
        this.authPass = authPass;
        return this;
    }

    public RestRequestBuilder nextWithAuthentication(String authUser, String authPass) {
        this.authUserNext = authUser;
        this.authPassNext = authPass;
        return this;
    }

    public RestRequestBuilder withHeader(String header, String value) {
        headers.put(header, value);
        return this;
    }

    public JerseyInvocation.Builder request(String path) {
        return addProperties(client.target(hostAnd(path))).request();
    }

    public JerseyInvocation.Builder request(String path, Map<String, String> routeParams) {
        for (Map.Entry<String, String> entry : routeParams.entrySet()) {
            path = path.replace("{"+entry.getKey()+"}", escaper.escape(entry.getValue()));
        }
        return request(path);
    }

    public String getHost() {
        return host;
    }

    public String getAuthUser() {
        return authUser;
    }

    public String getAuthPass() {
        return authPass;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    private JerseyWebTarget addProperties(JerseyWebTarget target) {
        for (Map.Entry<String, String> entry : headers.entrySet()) {
            target = target.property(entry.getKey(), entry.getValue());
        }
        if (authUserNext != null && authPassNext != null) {
            target = target.register(HttpAuthenticationFeature.basic(authUserNext, authPassNext));
            authUserNext = null;
            authPassNext = null;
        } else if (authUser != null && authPass != null) {
            target = target.register(HttpAuthenticationFeature.basic(authUser, authPass));
        }
        return target;
    }

    private String hostAnd(String path) {
        return host == null ? path : host + path;
    }
}

File: src/main/kotlin/com/meltwater/puppy/rest/RestRequestBuilder.kt
package com.meltwater.puppy.rest

import com.google.common.net.UrlEscapers
import org.glassfish.jersey.client.JerseyClientBuilder
import org.glassfish.jersey.client.JerseyInvocation
import org.glassfish.jersey.client.JerseyWebTarget
import org.glassfish.jersey.client.authentication.HttpAuthenticationFeature
import org.slf4j.LoggerFactory
import java.util.*
import java.util.Optional.empty

class RestRequestBuilder(var host: String, var auth: Pair<String, String>) {

    private val log = LoggerFactory.getLogger(RestRequestBuilder::class.java)

    private val client = JerseyClientBuilder.createClient()
    private val escaper = UrlEscapers.urlPathSegmentEscaper()

    private var authNext: Optional<Pair<String, String>> = empty()
    private var headers = HashMap<String, String>()

    fun nextWithAuthentication(authUser: String, authPass: String): RestRequestBuilder {
        authNext = Optional.of(Pair(authUser, authPass))
        return this
    }

    fun withHeader(header: String, value: String): RestRequestBuilder {
        headers.put(header, value)
        return this
    }

    fun request(path: String): JerseyInvocation.Builder {
        val url = hostAnd(path)
        log.trace("Building request to $url")
        return addProperties(client.target(url)).request()
    }

    fun request(path: String, routeParams: Map<String, String>): JerseyInvocation.Builder {
        var p: String = path
        for (entry in routeParams.entries) {
            p = p.replace("{${entry.key}}", escaper.escape(entry.value))
        }
        return request(p)
    }

    fun getAuthUser(): String = auth.first
    fun getAuthPass(): String = auth.second

    private fun addProperties(target: JerseyWebTarget): JerseyWebTarget {
        for (entry in headers.entries) {
            target.property(entry.key, entry.value)
        }
        if (authNext.isPresent) {
            target.register(HttpAuthenticationFeature.basic(authNext.get().first, authNext.get().second))
            authNext = empty()
        } else {
            target.register(HttpAuthenticationFeature.basic(auth.first, auth.second))
        }
        return target
    }

    private fun hostAnd(path: String): String = host + path
}
